src/Combinatorics/Matroids/matroids.jl

```diff
@@ -0,0 +1,821 @@
+export
+    Matroid, matroid_groundset, 
+    matroid_from_revlex_basis_encoding,
+    matroid_from_bases, matroid_from_nonbases, matroid_from_circuits, matroid_from_hyperplanes,
+    matroid_from_matrix_columns, matroid_from_matrix_rows,
+    cycle_matroid, bond_matroid, cocycle_matroid,
+    dual_matroid, direct_sum, restriction, deletion, contraction, minor,
+    principal_extension, free_extension, series_extension, parallel_extension,
+    uniform_matroid, fano_matroid, non_fano_matroid, pappus_matroid, non_pappus_matroid, vamos_matroid,
+    all_subsets_matroid, projective_plane, projective_geometry, affine_geometry
+
+
+################################################################################
+##  Constructing
+################################################################################
+ElementType = Union{IntegerUnion,Char,String}
+GroundsetType = Union{AbstractVector, AbstractSet} 
```

Comment:
Oh and definitely make these `const`, else this wreaks havoc on type stability and performance
```suggestion
const ElementType = Union{IntegerUnion,Char,String}
const GroundsetType = Union{AbstractVector, AbstractSet} 
```